src/Development/IDE/Core/Tracing.hs

```diff
@@ -0,0 +1,166 @@
+
+measureMemory :: [[Key]] -> (Maybe Key -> IO OurValueObserver) -> Var Values -> IO ()
+measureMemory groups instrumentFor stateRef = withSpan_ "Measure Memory" $ do
+    values <- readVar stateRef
+    valuesSizeRef <- newIORef 0
+    let !groupsOfGroupedValues = groupValues values
+    traceIO "STARTING MEMORY PROFILING"
+    forM_ groupsOfGroupedValues $ \groupedValues ->
+        repeatUntilJust $ do
+        traceIO (show $ map fst groupedValues)
+        runHeapsize $
+            forM_ groupedValues $ \(k,v) -> withSpan ("Measure " <> (BS.pack $ show k)) $ \sp -> do
+            acc <- liftIO $ newIORef 0
+            observe <- liftIO $ instrumentFor $ Just k
+            mapM_ (recursiveSize >=> \x -> liftIO (modifyIORef' acc (+ x))) v
+            size <- liftIO $ readIORef acc
+            let !byteSize = sizeOf (undefined :: Word) * size
+            setTag sp "size" (BS.pack (show byteSize ++ " bytes"))
+            () <- liftIO $ observe byteSize
+            liftIO $ modifyIORef' valuesSizeRef (+ byteSize)
+
+    valuesSize <- readIORef valuesSizeRef
+    observe <- instrumentFor Nothing
+    observe valuesSize
+    traceIO "MEMORY PROFILING COMPLETED"
+    where
+        blacklist = []
+
+        groupValues :: Values -> [ [(Key, [Value Dynamic])] ]
+        groupValues values =
+            let groupedValues =
+                    [ [ (k, vv)
+                      | k <- groupKeys
+                      , let vv = [ v | ((_,k'), v) <- HMap.toList values , k == k']
+                      ]
+                    | groupKeys <- groups
+                    ]
+                otherValues =
+                    HMap.toList $
+                    HMap.fromListWith (++)
+                    [ (k, [v])
+                    | ((_, k), v) <- HMap.toList values
+                    , k `notElem` blacklist
+                    , k `notElem` concat groups
+                    ]
+                !result = groupedValues <> map (:[]) otherValues
+                -- force the spine of the nested lists
+            in result `using` seqList (seqList (seqTuple2 r0 (seqList r0)))
+
+groupForObservableSharing :: [[Key]]
+groupForObservableSharing =
+        -- Group keys in bundles to be analysed jointly (for sharing)
+        -- Ideally all the keys would be analysed together, but if we try to do too much the GC will interrupt the analysis
+            [
+              [ Key GhcSessionDeps
+              , Key GhcSessionIO
+              , Key GhcSession
+              , Key TypeCheck
+              , Key GetModIface
+              , Key GetHieAst
+              , Key GetModSummary
+              , Key GetModSummaryWithoutTimestamps
+              , Key GetDependencyInformation
+              , Key GetModuleGraph
+              , Key GenerateCore
+              , Key GetParsedModule
+              ]
+            ]
+
+repeatUntilJust :: Monad m => m (Maybe b) -> m b
+repeatUntilJust action = do
```

Comment:
Maybe we want a sensible upper bound for the recursion here, so it doesn't repeat forever.

Reply:
Yes, indeed